File: sdks/python/settlekit/resources/coupons.py
"""Coupons resource — create, validate, and redeem discount coupons."""

from __future__ import annotations

from typing import Any, Mapping, Optional

from ._base import AsyncResource, SyncResource, clean_body


def percent_off(percent_off: float) -> dict[str, Any]:
    """Build a percent discount payload: ``{type: "percent", percentOff}``."""
    return {"type": "percent", "percentOff": percent_off}


def amount_off(amount_off: str) -> dict[str, Any]:
    """Build a fixed-amount discount payload: ``{type: "amount", amountOff}``."""
    return {"type": "amount", "amountOff": amount_off}


def free_trial_days(days: int) -> dict[str, Any]:
    """Build a free-trial discount payload: ``{type: "free-trial-days", days}``."""
    return {"type": "free-trial-days", "days": days}


def _create_body(
    *,
    code: str,
    discount: Mapping[str, Any],
    name: Optional[str],
    max_redemptions: Optional[int],
) -> dict[str, Any]:
    return clean_body(
        {
            "code": code,
            "discount": dict(discount),
            "name": name,
            "maxRedemptions": max_redemptions,
        }
    )


def _apply_body(
    *, subtotal: str, customer_id: Optional[str]
) -> dict[str, Any]:
    return clean_body({"subtotal": subtotal, "customerId": customer_id})


class Coupons(SyncResource):
    """Synchronous coupons namespace."""

    def create(
        self,
        *,
        code: str,
        discount: Mapping[str, Any],
        name: Optional[str] = None,
        max_redemptions: Optional[int] = None,
    ) -> Any:
        """Create a coupon. ``POST /v1/coupons``.

        Build ``discount`` with :func:`percent_off`, :func:`amount_off`, or
        :func:`free_trial_days`.
        """
        body = _create_body(
            code=code, discount=discount, name=name, max_redemptions=max_redemptions
        )
        return self._request("POST", "/v1/coupons", body=body)

    def list(self) -> Any:
        """List coupons. ``GET /v1/coupons``."""
        return self._request("GET", "/v1/coupons")

    def get(self, code: str) -> Any:
        """Fetch a coupon by code. ``GET /v1/coupons/:code``."""
        return self._request("GET", f"/v1/coupons/{code}")

    def validate(
        self, code: str, *, subtotal: str, customer_id: Optional[str] = None
    ) -> Any:
        """Validate a coupon against a subtotal. ``POST /v1/coupons/:code/validate``."""
        return self._request(
            "POST",
            f"/v1/coupons/{code}/validate",
            body=_apply_body(subtotal=subtotal, customer_id=customer_id),
        )

    def redeem(
        self, code: str, *, subtotal: str, customer_id: Optional[str] = None
    ) -> Any:
        """Redeem a coupon against a subtotal. ``POST /v1/coupons/:code/redeem``."""
        return self._request(
            "POST",
            f"/v1/coupons/{code}/redeem",
            body=_apply_body(subtotal=subtotal, customer_id=customer_id),
        )


class AsyncCoupons(AsyncResource):
    """Asynchronous coupons namespace."""

    async def create(
        self,
        *,
        code: str,
        discount: Mapping[str, Any],
        name: Optional[str] = None,
        max_redemptions: Optional[int] = None,
    ) -> Any:
        """Create a coupon. ``POST /v1/coupons``."""
        body = _create_body(
            code=code, discount=discount, name=name, max_redemptions=max_redemptions
        )
        return await self._request("POST", "/v1/coupons", body=body)

    async def list(self) -> Any:
        """List coupons. ``GET /v1/coupons``."""
        return await self._request("GET", "/v1/coupons")

    async def get(self, code: str) -> Any:
        """Fetch a coupon by code. ``GET /v1/coupons/:code``."""
        return await self._request("GET", f"/v1/coupons/{code}")

    async def validate(
        self, code: str, *, subtotal: str, customer_id: Optional[str] = None
    ) -> Any:
        """Validate a coupon. ``POST /v1/coupons/:code/validate``."""
        return await self._request(
            "POST",
            f"/v1/coupons/{code}/validate",
            body=_apply_body(subtotal=subtotal, customer_id=customer_id),
        )

    async def redeem(
        self, code: str, *, subtotal: str, customer_id: Optional[str] = None
    ) -> Any:
        """Redeem a coupon. ``POST /v1/coupons/:code/redeem``."""
        return await self._request(
            "POST",
            f"/v1/coupons/{code}/redeem",
            body=_apply_body(subtotal=subtotal, customer_id=customer_id),
        )

File: apps/cli/src/commands/coupons.ts
/**
 * `settlekit coupons` — manage discount coupons.
 *
 *   list                 GET  /v1/coupons
 *   create               POST /v1/coupons
 *   validate <code>      POST /v1/coupons/:code/validate
 *   redeem <code>        POST /v1/coupons/:code/redeem
 */
import type { Command } from "commander";
import { buildContext } from "../context.js";
import type { Money } from "../api.js";

interface Coupon extends Record<string, unknown> {
  code: string;
  status: string;
  redeemedCount: number;
}

/** Build the discount payload from the --percent-off / --amount-off / --free-trial-days flags. */
function buildDiscount(opts: Record<string, unknown>): Record<string, unknown> {
  if (opts.percentOff !== undefined) return { type: "percent", percentOff: opts.percentOff };
  if (opts.amountOff !== undefined) return { type: "amount", amountOff: opts.amountOff };
  if (opts.freeTrialDays !== undefined) return { type: "free-trial-days", days: opts.freeTrialDays };
  throw new Error("Provide one of --percent-off, --amount-off, or --free-trial-days.");
}

export function registerCoupons(program: Command): void {
  const coupons = program.command("coupons").description("Manage discount coupons");

  coupons
    .command("list")
    .description("List coupons")
    .action(async function (this: Command) {
      const ctx = buildContext(this);
      const rows = await ctx.client.get<Coupon[]>("/v1/coupons");
      ctx.printList(rows, [
        { header: "CODE", value: (c) => c.code },
        { header: "STATUS", value: (c) => c.status },
        { header: "REDEEMED", value: (c) => c.redeemedCount },
      ]);
    });

  coupons
    .command("create")
    .description("Create a coupon")
    .requiredOption("--code <code>", "Coupon code")
    .option("--name <name>", "Display name")
    .option("--percent-off <n>", "Percent discount (1-100)", (v) => Number.parseInt(v, 10))
    .option("--amount-off <amount>", "Fixed amount off, e.g. 5.00")
    .option("--free-trial-days <n>", "Free trial days", (v) => Number.parseInt(v, 10))
    .option("--max-redemptions <n>", "Global redemption cap", (v) => Number.parseInt(v, 10))
    .action(async function (this: Command) {
      const opts = this.opts();
      const ctx = buildContext(this);
      const body: Record<string, unknown> = { code: opts.code, discount: buildDiscount(opts) };
      if (opts.name) body.name = opts.name;
      if (opts.maxRedemptions !== undefined) body.maxRedemptions = opts.maxRedemptions;
      const coupon = await ctx.client.post<Coupon>("/v1/coupons", body);
      ctx.printRecord(coupon);
    });

  coupons
    .command("validate <code>")
    .description("Dry-run apply a coupon to a subtotal")
    .requiredOption("--subtotal <amount>", "Subtotal, e.g. 100.00")
    .option("--customer-id <id>", "Customer id (for per-customer limits)")
    .action(async function (this: Command, code: string) {
      const opts = this.opts();
      const ctx = buildContext(this);
      const body: Record<string, unknown> = { subtotal: opts.subtotal };
      if (opts.customerId) body.customerId = opts.customerId;
      const result = await ctx.client.post<{ ok: boolean; discount: Money; total: Money }>(
        `/v1/coupons/${encodeURIComponent(code)}/validate`,
        body,
      );
      ctx.printRecord(result as unknown as Record<string, unknown>);
    });

  coupons
    .command("redeem <code>")
    .description("Redeem a coupon against a subtotal")
    .requiredOption("--subtotal <amount>", "Subtotal, e.g. 100.00")
    .option("--customer-id <id>", "Customer id")
    .action(async function (this: Command, code: string) {
      const opts = this.opts();
      const ctx = buildContext(this);
      const body: Record<string, unknown> = { subtotal: opts.subtotal };
      if (opts.customerId) body.customerId = opts.customerId;
      const result = await ctx.client.post<Record<string, unknown>>(
        `/v1/coupons/${encodeURIComponent(code)}/redeem`,
        body,
      );
      ctx.printRecord(result);
    });
}
